README.md
# pyFirmata2

![alt tag](screenshot.png)

PyFirmata2 turns your Arduino into a data acquisition card controlled by Python.

Up to 100Hz precise sampling at the analogue ports for digital filtering.

Just upload the default firmata sketch into your Arduino and you are all set.

pyFirmata2 is an updated version of pyFirmata which adds *precise sampling of the analogue inputs*
to the API so that it's possible to filter signals and in general do
signal processing. Instead of "sleep" commands which have unreliable timing
the Arduino performs the sampling in its firmware and transmits the data
then to pyFirmata2. The Python application simply registers a callback
which is then called every time after new data has arrived.

This API has been used in my Digital Signal Processing (DSP) class to
practise realtime filtering of analogue sensor
data. Examples can be viewed on the YouTube channel of the
class: https://www.youtube.com/user/DSPcourse


## Installation



### Upload firmata


Install the Arduino IDE on your computer: https://www.arduino.cc/en/Main/Software

Start the IDE and upload the standard firmata sketch into your Arduino with:
```  
    File -> Examples -> Firmata -> Standard Firmata
```


### Install pyfirmata2


The preferred way to install is with `pip` / `pip3`. Under Linux:
```
    pip3 install pyfirmata2 [--user] [--upgrade]
```
    
and under Windows/Mac type:
```  
    pip install pyfirmata2 [--user] [--upgrade]
```
    
You can also install from source with:
```
    git clone https://github.com/berndporr/pyFirmata2
    cd pyFirmata2
```

Under Linux type:
```  
    python3 setup.py install
```

Under Windows / Mac:
```
    python setup.py install
```

## Usage


### Initialisation

Create an instance of the `Arduino` class:
```
    from pyfirmata import Arduino
    board = Arduino(Arduino.AUTODETECT)
```
which automatically detects the serial port of the Arduino.

If this fails you can also specify the serial port manually, for example:
```
    board = Arduino('COM4')
```
Under Linux this is usually `/dev/ttyACM0`. Under Windows this is a
COM port, for example `COM4`. On a MAC it's `/dev/ttys000`, `/dev/cu.usbmodem14101` or
check for the latest addition: `ls -l -t /dev/*`.


### Starting sampling at a given sampling interval

In order to sample analogue data you need to specify a
sampling interval in ms. The smallest interval is 10ms:
```
    board.samplingOn(samplinginterval in ms)
```
Calling `samplingOn()` without its argument sets
the sampling interval to 19ms.


### Enabling and reading from analoge pins

To process data at a given sampling interval register a callback
handler and then enable it:
```
    board.analog[0].register_callback(myCallback)
    board.analog[0].enable_reporting()
```    
where `myCallback(data)` is then called every time after data has been received
and is timed by the arduino itself.

You can also read additional analogue pins any time by issuing a read
command:
```
    board.analog[1].read()
```
This is useful for reading additional pins within a callback handler
to process multiple pins simultaneously. Note that the data obtained
by `read()` is read from an internal buffer which stores the most
recent value received from the Arduino. This call is non-blocking.
You also need to run `enable_reporting()` on that pin before you can use `read()`.


### Writing to a digital port

Digital ports can be written to at any time:
```  
    board.digital[13].write(1)
```
For any other functionality use the pin class below.

    
### The pin class

The command `get_pin` requests the class of a pin
by specifying a string, composed of
'a' or 'd' (depending on if you need an analog or digital pin), the pin
number, and the mode ('i' for input, 'o' for output, 'p' for pwm). All
seperated by `:`. Eg. `a:0:i` for analog 0 as input or `d:3:p` for
digital pin 3 as pwm:
```
    analog_0 = board.get_pin('a:0:i')
    analog_0.read()
    pin3 = board.get_pin('d:3:p')
    pin3.write(0.6)
```	
	
### Closing the board

To close the serial port to the Arduino use the exit command:
```    
	board.exit()
```

## Example code

The directory https://github.com/berndporr/pyFirmata2/tree/master/examples 
contains two realtime Oscilloscopes with precise sampling rate,
a digital port reader, the ubiquitous flashing LED program and
a program which prints data using the callback handler.


## Troubleshooting

### Spyder

Start your program from the (Anaconda-) console / terminal and never within Spyder. Here is
an example for Windows::
```
    (base) D:\>
    (base) D:\>cd pyFirmata2\examples
    (base) D:\pyFirmata2\examples>python realtime_two_channel_scope.py
```
The problem with Spyder is that it won't let your Python program terminate properly
which leaves the serial port in an undefined state. If you then re-run your program
it won't be able to talk to your Arduino. In the worst case you need to reboot your
computer. Bottomline: use Spyder for editing, run the program from the console / terminal.


### After an update still the old version is being used

If you use the `--user` option to install / update packages Python might keep older versions.

Solution: Do a `pip uninstall pyfirmata2` multiple times until no version is left 
on your computer. Then install it again as described above.




### Credits

The original pyFirmata was written by Tino de Bruijn.
The realtime sampling / callback has been added by Bernd Porr.


CHANGES.rst
=======
Changes
=======

Version 1.1.x
=============

1.1.0 - 10/03/2019
------------------

- Nano board added
- Iterator daemon now exits properly

Version 1.0.x
=============

1.0.3 - 17/05/2015
------------------

- Pass ``timeout`` parameter on board trough to ``pyserial.Serial``.
- Added this change list.

1.0.2 - 17/01/2015
------------------

- Configure ``bumpversion``.
- Update to-do list.

1.0.1 - 17/01/2015
------------------

- Added Firmata's "Capability Query", that allows to auto setup a board. (This probably deserved a minor version bump...)
- Start distributing as wheels

1.0.0 - 04/01/2015
------------------

- Added Python 3 support
- Testing on Python 2.6 is dropped, but it might still work. Not actively supported though.


pyfirmata/util.py
from __future__ import division, unicode_literals

import os
import sys
import threading
import time

import serial

from .boards import BOARDS


def get_the_board(
    layout=BOARDS["arduino"], base_dir="/dev/", identifier="tty.usbserial"
):
    """
    Helper function to get the one and only board connected to the computer
    running this. It assumes a normal arduino layout, but this can be
    overriden by passing a different layout dict as the ``layout`` parameter.
    ``base_dir`` and ``identifier`` are overridable as well. It will raise an
    IOError if it can't find a board, on a serial, or if it finds more than
    one.
    """
    from .pyfirmata import Board  # prevent a circular import

    boards = []
    for device in os.listdir(base_dir):
        if device.startswith(identifier):
            try:
                board = Board(os.path.join(base_dir, device), layout)
            except serial.SerialException:
                pass
            else:
                boards.append(board)
    if len(boards) == 0:
        raise IOError(
            "No boards found in {0} with identifier {1}".format(base_dir, identifier)
        )
    elif len(boards) > 1:
        raise IOError("More than one board found!")
    return boards[0]


class Iterator(threading.Thread):

    def __init__(self, board):
        super(Iterator, self).__init__()
        self.board = board
        self.daemon = True

    def run(self):
        while 1:
            try:
                while self.board.bytes_available():
                    self.board.iterate()
                time.sleep(0.001)
            except (AttributeError, serial.SerialException, OSError):
                # this way we can kill the thread by setting the board object
                # to None, or when the serial port is closed by board.exit()
                break
            except Exception as e:
                # catch 'error: Bad file descriptor'
                # iterate may be called while the serial port is being closed,
                # causing an "error: (9, 'Bad file descriptor')"
                if getattr(e, "errno", None) == 9:
                    break
                try:
                    if e[0] == 9:
                        break
                except (TypeError, IndexError):
                    pass
                raise
            except (KeyboardInterrupt):
                sys.exit()


def to_two_bytes(integer):
    """
    Breaks an integer into two 7 bit bytes.
    """
    if integer > 32767:
        raise ValueError("Can't handle values bigger than 32767 (max for 2 bits)")
    return bytearray([integer % 128, integer >> 7])


def from_two_bytes(bytes):
    """
    Return an integer from two 7 bit bytes.
    """
    lsb, msb = bytes
    try:
        # Usually bytes have been converted to integers with ord already
        return msb << 7 | lsb
    except TypeError:
        # But add this for easy testing
        # One of them can be a string, or both
        try:
            lsb = ord(lsb)
        except TypeError:
            pass
        try:
            msb = ord(msb)
        except TypeError:
            pass
        return msb << 7 | lsb


def two_byte_iter_to_str(bytes):
    """
    Return a string made from a list of two byte chars.
    """
    bytes = list(bytes)
    chars = bytearray()
    while bytes:
        lsb = bytes.pop(0)
        try:
            msb = bytes.pop(0)
        except IndexError:
            msb = 0x00
        chars.append(from_two_bytes([lsb, msb]))
    return chars.decode()


def str_to_two_byte_iter(string):
    """
    Return a iter consisting of two byte chars from a string.
    """
    bstring = string.encode()
    bytes = bytearray()
    for char in bstring:
        bytes.append(char)
        bytes.append(0)
    return bytes


def break_to_bytes(value):
    """
    Breaks a value into values of less than 255 that form value when multiplied.
    (Or almost do so with primes)
    Returns a tuple
    """
    if value < 256:
        return (value,)
    c = 256
    least = (0, 255)
    for i in range(254):
        c -= 1
        rest = value % c
        if rest == 0 and value / c < 256:
            return (c, int(value / c))
        elif rest == 0 and value / c > 255:
            parts = list(break_to_bytes(value / c))
            parts.insert(0, c)
            return tuple(parts)
        else:
            if rest < least[1]:
                least = (c, rest)
    return (c, int(value / c))


def pin_list_to_board_dict(pinlist):
    """
    Capability Response codes:
        INPUT:  0, 1
        OUTPUT: 1, 1
        ANALOG: 2, 10
        PWM:    3, 8
        SERV0:  4, 14
        I2C:    6, 1
    """

    board_dict = {
        "digital": [],
        "analog": [],
        "pwm": [],
        "servo": [],  # 2.2 specs
        # 'i2c': [],  # 2.3 specs
        "disabled": [],
    }
    for i, pin in enumerate(pinlist):
        pin.pop()  # removes the 0x79 on end
        if not pin:
            board_dict["disabled"] += [i]
            board_dict["digital"] += [i]
            continue

        for j, _ in enumerate(pin):
            # Iterate over evens
            if j % 2 == 0:
                # This is safe. try: range(10)[5:50]
                if pin[j:j + 4] == [0, 1, 1, 1]:
                    board_dict["digital"] += [i]

                if pin[j:j + 2] == [2, 10]:
                    board_dict["analog"] += [i]

                if pin[j:j + 2] == [3, 8]:
                    board_dict["pwm"] += [i]

                if pin[j:j + 2] == [4, 14]:
                    board_dict["servo"] += [i]

                # Desable I2C
                if pin[j:j + 2] == [6, 1]:
                    pass

    # We have to deal with analog pins:
    # - (14, 15, 16, 17, 18, 19)
    # + (0, 1, 2, 3, 4, 5)
    diff = set(board_dict["digital"]) - set(board_dict["analog"])
    board_dict["analog"] = [n for n, _ in enumerate(board_dict["analog"])]

    # Digital pin problems:
    # - (2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15)
    # + (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13)

    board_dict["digital"] = [n for n, _ in enumerate(diff)]
    # Based on lib Arduino 0017
    board_dict["servo"] = board_dict["digital"]

    # Turn lists into tuples
    # Using dict for Python 2.6 compatibility
    board_dict = dict([(key, tuple(value)) for key, value in board_dict.items()])

    return board_dict
